tools/debuggers.md

```diff
@@ -0,0 +1,86 @@
+
+1. **Understand what your code _should_ do**: It's hard to solve a problem if you don't know when you've found a solution! Understanding a program's expected behavior is what code _specifications_ and _tests_ are for, head over to those chapters of the Study Book to learn more about understanding code's expected behavior.
+2. **Understand what your code _actually_ does**: This is usually simple enough, run the code and see what happens! Does the button turn the wrong color when you click it? Is there an error logged to the console?
+3. **Find the problem line(s) of code**: This is the most important part! If you can't figure out exactly which line(s) of code are causing problems, your best hope is to randomly change things. Random changes is a recipe to make things worse :( Inspecting program memory and stepping through your code one line at a time are necessary skills for discovering which line(s) of code are causing your program to break.
+4. **Fix the problem line(s) of code**: This step is very difficult and can take a long time to master. Becoming a master debugger will take many hours of: reading other people's code to learn what good code looks like, collecting different strategies for different situations, and lots and lots of practice.
+
+So let's take a look at some tools available to help you with #3, tools that will help you step through your code and understand what is happening in memory:
+
+## Console Logs
+
+Using `console.log` is the simplest debugging tool available. When you console.log a variable, you're printing the value that is stored in memory. If that value is different than what you expected then it's time to start debugging (careful though! sometimes the value is what you expect, but you expected the wrong thing ;)
+
+As you explore your code you can remove and add logs at different lines to narrow down which part of your program is confusing.
+
+- **Pros**:
+  - it's easy to add or remove logs
+  - works in Node, the Browser, JS Tutor, and any other JS environment
+  - you can focus in on a few lines of code without being distracted by everything else in memory (unlike the other debugger tools)
+- **Cons**:
+  - you will only find what you look for. If you don't think of logging a variable, then you cannot see what it stores. Very often mistakes occur where you don't expect them to!
+  - logs will give an incomplete idea of your code's execution, it's always possible to figure out which line of code was executed before your log or after. Often this is the information you need to fix your problem.
+  - console logs can not give you a full picture of how program memory is organized, and how it changes over time in your program. For this reason, logging is often most helpful with code you mostly understand instead of code you mostly _don't_ understand.
+
+## JS Tutor
```

Comment:
This is not really a debugger, is it? It is more a learning tool as indicated by its name.

Reply:
A lot of our students (especially the best) use it to debug their projects by refactoring out the code they don't understand, debugging it there, then pasting it back in.  

i suppose that workflow could be made more clear

Reply:
Is that because they don't know how to use the debugger? I guess you wouldn't want them to put experience with JSTutor on their CV 😁.

Reply:
It's because it helps to learn the debugger, students in Be started using the devtools much more effectively after we introduced JS Tutor first

and I also don't see the harm bringing up JS Tutor in an interview, if it helps you write better code then it's a tool your colleagues should be happy you know

Reply:
I agree that it is not really a debugger, but do like it in this list as something students may find useful to understand code. Maybe we can make it clearer by adjusting the text:

`Besides being a debugging tool to for understanding code that does not do what you expect, it's also a great tool for gaining a deeper knowledge of code you already do understand.`

to:
`Although technically not a debugger, JS Tutor does help gain a deeper knowledge of your code so it is included here.`